src/solvers/refinement/string_refinement.cpp

```diff
@@ -473,27 +470,16 @@ void string_refinementt::set_to(const exprt &expr, bool value)
       }
     }
 
-    if(value)
-    {
-      if(!add_axioms_for_string_assigns(lhs, subst_rhs))
-        return;
-    }
-    else
-    {
-      // TODO: Something should also be done if value is false.
-      INVARIANT(
-        !is_char_array(eq_expr.rhs().type()),
-        string_refinement_invariantt("set_to cannot set a char_array"));
-      INVARIANT(
-        !refined_string_typet::is_refined_string_type(eq_expr.rhs().type()),
-        string_refinement_invariantt("set_to cannot set a refined_string"));
-    }
+    if(value && !add_axioms_for_string_assigns(lhs, subst_rhs))
+      return;
 
+    // Push the substituted equality to the list of axioms to be given to
+    // supert::set_to.
     non_string_axioms.push_back(std::make_pair(equal_exprt(lhs, subst_rhs),
                                                value));
   }
-  // We keep a list of the axioms to give to supert::set_to in order to
-  // substitute the symbols in dec_solve().
+  // Push the unmodified equality to the list of axioms to be given to
+  // supert::set_to.
   else
     non_string_axioms.push_back(std::make_pair(expr, value));
```

Comment:
maybe we should substitute string function in the expression even if it is not an equality, or better assert that there is no refined strings or char array in that expression